assets/js/googlesitekit/modules/create-submit-changes-store.js

```diff
@@ -0,0 +1,201 @@
+export const INVARIANT_SETTINGS_NOT_CHANGED = 'cannot submit changes if settings have not changed';
+
+// Actions
+const SUBMIT_CHANGES = 'SUBMIT_CHANGES';
+const START_SUBMIT_CHANGES = 'START_SUBMIT_CHANGES';
+const FINISH_SUBMIT_CHANGES = 'FINISH_SUBMIT_CHANGES';
+
+/**
+ * Creates a store object implementing the necessary infrastructure for submitting module settings.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Object}   args                            Arguments for creating the submitChanges store.
+ * @param {string}   args.storeName                  Datastore slug.
+ * @param {Function} [args.submitChanges]            Optional. Callback function to issue the submit changes request. Will be used inside the submit changes control.
+ * @param {Function} [args.validateCanSubmitChanges] Optional. A helper function to validate that settings can be submitted.
+ * @return {Object} Partial store object with properties 'actions', 'controls', 'reducer', 'resolvers', and 'selectors'.
+ */
+export function createSubmitChangesStore( {
+	storeName,
+	submitChanges = () => () => ( {} ),
+	validateCanSubmitChanges = () => {},
+} = {} ) {
+	invariant( storeName, 'storeName is required.' );
+
+	const initialState = {
+		hasStartedSubmittingChanges: false,
+		hasFinishedSubmittingChanges: false,
+		isDoingSubmitChanges: false,
+	};
+
+	const actions = {
+		/**
+		 * Submits all changes currently present in the client, persisting them on the server.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @return {Object} Empty object on success, object with `error` property on failure.
+		 */
+		*submitChanges() {
+			const { dispatch } = yield Data.commonActions.getRegistry();
+			dispatch( storeName ).clearError( 'submitChanges', [] );
+
+			yield {
+				type: START_SUBMIT_CHANGES,
+				payload: {},
+			};
+
+			const result = yield {
+				type: SUBMIT_CHANGES,
+				payload: {},
+			};
+
+			if ( result.error ) {
+				yield dispatch( storeName ).receiveError( result.error, 'submitChanges', [] );
+			}
+
+			yield {
+				type: FINISH_SUBMIT_CHANGES,
+				payload: {},
+			};
+
+			return result;
+		},
+	};
+
+	const reducer = ( state, { type } ) => {
+		switch ( type ) {
+			case START_SUBMIT_CHANGES: {
+				return {
+					...state,
+					hasStartedSubmittingChanges: true,
+					hasFinishedSubmittingChanges: false,
+					isDoingSubmitChanges: true,
+				};
+			}
+
+			case FINISH_SUBMIT_CHANGES: {
+				return {
+					...state,
+					hasStartedSubmittingChanges: false,
+					hasFinishedSubmittingChanges: true,
+					isDoingSubmitChanges: false,
+				};
+			}
+
+			default: {
+				return state;
+			}
+		}
+	};
+
+	const controls = {
+		[ SUBMIT_CHANGES ]: submitChanges,
+	};
+
+	const resolvers = {};
+
+	const {
+		safeSelector: canSubmitChanges,
+		dangerousSelector: validateSubmitChanges,
+	} = createValidationSelector( validateCanSubmitChanges );
+
+	const selectors = {
+		canSubmitChanges,
+		validateSubmitChanges,
+		...createStateSelectors( Object.keys( initialState ) ),
```

Comment:
I think this abstraction goes a bit too far - especially since we only need a single state property, and for the benefit of better documentation, let's write out the selector `isDoingSubmitChanges` instead of doing this.

Reply:
Updated.